src/benchmark.py

```diff
@@ -0,0 +1,201 @@
+        asyncio.run(benchmark_remote(config=S3ConfigBase(ceph), bucket_id=bucket_id))
+
+
+async def benchmark_remote(config: S3ConfigBase, bucket_id=str):
+    """Run against a remote endpoint based on the given config"""
+    WithRetry.set_retries(3)
+    object_storage = S3ObjectStorage(config=config)
+    await benchmark_upload(object_storage=object_storage, bucket_id=bucket_id)
+    await benchmark_download(object_storage=object_storage, bucket_id=bucket_id)
+
+
+async def benchmark_localstack(bucket_id: str):
+    """Create bucket and run up-/download benchmarks"""
+    # assume localstack should be fairly reliable
+    WithRetry.set_retries(0)
+    with LocalStackContainer(image="localstack/localstack:0.14.2").with_services(
+        "s3"
+    ) as localstack:
+        config = config_from_localstack_container(localstack)
+        storage = S3ObjectStorage(config=config)
+        await storage.create_bucket(bucket_id)
+        await benchmark_upload(object_storage=storage, bucket_id=bucket_id)
+        await benchmark_download(object_storage=storage, bucket_id=bucket_id)
+        await storage.delete_bucket(bucket_id)
+
+
+async def benchmark_upload(object_storage: S3ObjectStorage, bucket_id: str):
+    """Call and time actual upload per file"""
+    for path in FILE_PATHS:
+        print(f"Uploading file {path}")
+        upload_start = time.time()
+        await upload_object(
+            object_storage=object_storage, bucket_id=bucket_id, path=path
+        )
+        elapsed = time.time() - upload_start
+        print(f"Upload for file {path} finished in {elapsed:.2f}s")
+
+
+async def upload_object(object_storage: S3ObjectStorage, bucket_id: str, path: Path):
+    """Run and time upload of all parts"""
+    object_id = os.path.basename(path)
+    upload_id = await object_storage.init_multipart_upload(
+        bucket_id=bucket_id, object_id=object_id
+    )
+
+    upload_start = time.time()
+    total_parts = 0
+    try:
+        with open(path, "r+b") as source:
+            for (part_number, file_part) in enumerate(
+                read_file_parts(source, part_size=PART_SIZE), start=1
+            ):
+                part_upload_url = await object_storage.get_part_upload_url(
+                    upload_id=upload_id,
+                    bucket_id=bucket_id,
+                    object_id=object_id,
+                    part_number=part_number,
+                )
+                upload_file_part(presigned_url=part_upload_url, part=file_part)
+
+                duration = time.time() - upload_start
+                average = (PART_SIZE / 1024**2) / (duration / part_number)
+                print(
+                    f"\rAverage transfer rate: {average:.2f} MiB/s (Part number {part_number})",
+                    end="",
+                )
+                total_parts = part_number
+    except Exception as exc:  # pylint: disable=bare-except
+        # clean up multipart upload for next try, if we run into issues
```

Comment:
Wasn't miltipart upload cleaned up?
In that case we could bump the hexkit version number here and remove this line

Reply:
Hexkit version is bumped already.
This is to abort the multipart upload in case something else went wrong during the upload
If I just run the script in a loop in the commandline this is easier to deal with this way.
Also removes manual steps if you abort the script in the commandline with ctrl + c